src/software/ai/hl/stp/tactic/shadow_enemy/shadow_enemy_fsm.h

```diff
@@ -0,0 +1,219 @@
+#pragma once
+
+#include "software/ai/evaluation/calc_best_shot.h"
+#include "software/ai/evaluation/enemy_threat.h"
+#include "software/ai/hl/stp/tactic/move/move_fsm.h"
+#include "software/ai/hl/stp/tactic/tactic.h"
+#include "software/geom/algorithms/distance.h"
+
+struct ShadowEnemyFSM
+{
+    class BlockShotState;
+    class BlockPassState;
+    class StealAndChipState;
+
+    // this struct defines the unique control parameters that the ShadowEnemyFSM requires
+    // in its update
+    struct ControlParams
+    {
+        // The Enemy Threat indicating which enemy to shadow
+        std::optional<EnemyThreat> enemy_threat;
```

Comment:
I'm curious about this std::optional here. If there's no threat then the shadow enemy fsm could be "done"/dormant, but there isn't the logic to support transitioning to done on that basis

Reply:
I added support for when `enemy_threat = std:nullopt`, in both `block_pass` and `block_shot`, if no `enemy_threat` is detected, it will position itself `shadow_distance` away from the ball in the direction of our `friendlyGoalArea`. It essentially defaults to blocking the shot